scipy/interpolate/tests/test_aaa.py

```diff
@@ -0,0 +1,202 @@
+# Copyright (c) 2017, The Chancellor, Masters and Scholars of the University
+# of Oxford, and the Chebfun Developers. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#     * Redistributions of source code must retain the above copyright
+#       notice, this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above copyright
+#       notice, this list of conditions and the following disclaimer in the
+#       documentation and/or other materials provided with the distribution.
+#     * Neither the name of the University of Oxford nor the names of its
+#       contributors may be used to endorse or promote products derived from
+#       this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+import numpy as np
+from numpy.testing import assert_allclose, assert_equal
+import pytest
+import scipy
+from scipy.interpolate import AAA
+
+TOL = 1e4 * np.finfo(np.float64).eps
+UNIT_INTERVAL = np.linspace(-1, 1, num=1000)
+PTS = np.logspace(-15, 0, base=10, num=500)
+PTS = np.concatenate([-PTS[::-1], [0], PTS])
+
+class TestAAA:
+    def test_input_validation(self):
+        with pytest.raises(ValueError, match="same size"):
+            AAA([0], [1, 1])
+        with pytest.raises(ValueError, match="1-D"):
+            AAA([[0], [0]], [[1], [1]])
+    
+    def test_convergence_error(self):
+        with pytest.warns(RuntimeWarning, match="AAA failed"):
+            AAA(UNIT_INTERVAL, np.exp(UNIT_INTERVAL),  max_terms=1)
+
+    @pytest.mark.parametrize("dtype", [np.float32, np.float64, np.complex64, 
+                                       np.complex128])
+    def test_dtype_preservation(self, dtype):
+        z = np.linspace(-1, 1, dtype=dtype)
+        assert AAA(z, np.sin(z))(z).dtype == dtype
```

Comment:
Consider testing for accuracy, too. This is just the sort of test we needed in `differential_entropy` recently.

```suggestion
    def test_dtypes(self, dtype):
        rtol = np.finfo(dtype).eps ** 0.75 * 100
        rng = np.random.default_rng(59846294526092468)

        z = np.linspace(-1, 1, dtype=dtype)
        f = np.sin(z)
        aaa = AAA(z, f)

        z2 = (2*rng.random(size=100) - 1).astype(dtype)
        assert_allclose(aaa(z2), np.sin(z2), rtol=rtol)
        assert aaa(z2).dtype == dtype
```